File: donors/orchestration-java/CircuitBreaker.java
package com.aiassistant.core/orchestration;

import android.util.Log;

import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

public class CircuitBreaker {
    private static final String TAG = "CircuitBreaker";
    
    private final String componentId;
    private final int failureThreshold;
    private final long cooldownPeriodMs;
    
    private final AtomicInteger failureCount;
    private final AtomicLong lastFailureTime;
    private final AtomicInteger executionCount;
    
    private volatile State state;
    
    public CircuitBreaker(String componentId, int failureThreshold, long cooldownPeriodMs) {
        this.componentId = componentId;
        this.failureThreshold = failureThreshold;
        this.cooldownPeriodMs = cooldownPeriodMs;
        this.failureCount = new AtomicInteger(0);
        this.lastFailureTime = new AtomicLong(0);
        this.executionCount = new AtomicInteger(0);
        this.state = State.CLOSED;
    }
    
    public boolean allowExecution() {
        executionCount.incrementAndGet();
        
        if (state == State.OPEN) {
            long currentTime = System.currentTimeMillis();
            long timeSinceFailure = currentTime - lastFailureTime.get();
            
            if (timeSinceFailure >= cooldownPeriodMs) {
                Log.i(TAG, "Circuit breaker for " + componentId + 
                      " entering HALF_OPEN state after cooldown");
                state = State.HALF_OPEN;
                return true;
            }
            
            Log.w(TAG, "Circuit breaker for " + componentId + " is OPEN - blocking execution");
            return false;
        }
        
        return true;
    }
    
    public void recordSuccess() {
        failureCount.set(0);
        
        if (state == State.HALF_OPEN) {
            Log.i(TAG, "Circuit breaker for " + componentId + " transitioning to CLOSED");
            state = State.CLOSED;
        }
    }
    
    public void recordFailure() {
        int failures = failureCount.incrementAndGet();
        lastFailureTime.set(System.currentTimeMillis());
        
        if (failures >= failureThreshold && state != State.OPEN) {
            Log.w(TAG, "Circuit breaker for " + componentId + 
                  " OPEN - threshold reached (" + failures + "/" + failureThreshold + ")");
            state = State.OPEN;
        }
    }
    
    public void reset() {
        failureCount.set(0);
        state = State.CLOSED;
        Log.i(TAG, "Circuit breaker for " + componentId + " reset to CLOSED");
    }
    
    public State getState() {
        return state;
    }
    
    public int getFailureCount() {
        return failureCount.get();
    }
    
    public int getExecutionCount() {
        return executionCount.get();
    }
    
    public enum State {
        CLOSED,
        OPEN,
        HALF_OPEN
    }
}

File: android/app/src/main/kotlin/com/ariaagent/mobile/core/orchestration/CircuitBreaker.kt
// Ported from donors/orchestration-java/CircuitBreaker.java
// Original repo: TITANICBHAI/AI-ASSISTANT-INCOMPLETE
// Changes: fixed donor package typo (`com.aiassistant.core/orchestration`),
// idiomatic Kotlin with @Volatile state and AtomicInteger/AtomicLong counters.

package com.ariaagent.mobile.core.orchestration

import android.util.Log
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * Classic three-state circuit breaker (CLOSED → OPEN → HALF_OPEN → CLOSED).
 *
 * Wraps every call into a downstream component:
 *   1. CLOSED   — calls flow through; failures increment the counter.
 *   2. OPEN     — calls are short-circuited until [cooldownPeriodMs] elapses.
 *   3. HALF_OPEN — first call after cooldown is allowed; success → CLOSED,
 *                  failure → OPEN again.
 *
 * Thread-safe via atomics and a single @Volatile state. No locking required.
 */
class CircuitBreaker(
    val componentId: String,
    private val failureThreshold: Int,
    private val cooldownPeriodMs: Long,
) {
    private val failureCount = AtomicInteger(0)
    private val lastFailureTime = AtomicLong(0L)
    private val executionCount = AtomicInteger(0)

    @Volatile var state: State = State.CLOSED
        private set

    /** @return true if the call is allowed to proceed. */
    fun allowExecution(): Boolean {
        executionCount.incrementAndGet()
        if (state != State.OPEN) return true

        val timeSinceFailure = System.currentTimeMillis() - lastFailureTime.get()
        if (timeSinceFailure >= cooldownPeriodMs) {
            Log.i(TAG, "Circuit for $componentId entering HALF_OPEN after cooldown")
            state = State.HALF_OPEN
            return true
        }
        Log.w(TAG, "Circuit for $componentId is OPEN — blocking execution")
        return false
    }

    fun recordSuccess() {
        failureCount.set(0)
        if (state == State.HALF_OPEN) {
            Log.i(TAG, "Circuit for $componentId transitioning HALF_OPEN → CLOSED")
            state = State.CLOSED
        }
    }

    fun recordFailure() {
        val failures = failureCount.incrementAndGet()
        lastFailureTime.set(System.currentTimeMillis())
        if (failures >= failureThreshold && state != State.OPEN) {
            Log.w(TAG, "Circuit for $componentId OPEN — threshold reached ($failures/$failureThreshold)")
            state = State.OPEN
        }
    }

    fun reset() {
        failureCount.set(0)
        state = State.CLOSED
        Log.i(TAG, "Circuit for $componentId reset to CLOSED")
    }

    fun getFailureCount(): Int = failureCount.get()
    fun getExecutionCount(): Int = executionCount.get()

    enum class State { CLOSED, OPEN, HALF_OPEN }

    companion object {
        private const val TAG = "CircuitBreaker"
    }
}
